sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotExposeListT.cs

```diff
@@ -0,0 +1,150 @@
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+            {
+                if (c.IsTest())
+                {
+                    return;
+                }
+
+                var baseMethodDeclaration = (BaseMethodDeclarationSyntax)c.Node;
+                var methodSymbol = c.SemanticModel.GetDeclaredSymbol(baseMethodDeclaration);
+
+                if (methodSymbol == null ||
+                    !IsPubliclyAccessible(methodSymbol) ||
+                    methodSymbol.IsOverride ||
+                    IsPropertyGetOrSet(methodSymbol))
+                {
+                    return;
+                }
+
+                string methodType = methodSymbol.IsConstructor() ? "constructor" : "method";
+
+                var methodDeclaration = baseMethodDeclaration as MethodDeclarationSyntax;
+                if (methodDeclaration != null)
+                {
+                    ReportIfListT(methodDeclaration.ReturnType, c, methodType);
+                }
+
+                baseMethodDeclaration
+                    .ParameterList?
+                    .Parameters
+                    .ToList()
+                    .ForEach(p => ReportIfListT(p.Type, c, methodType));
+            },
+            SyntaxKind.MethodDeclaration,
+            SyntaxKind.ConstructorDeclaration);
+
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+            {
+                if (c.IsTest())
+                {
+                    return;
+                }
+
+                var propertyDeclaration = (PropertyDeclarationSyntax)c.Node;
+                var propertySymbol = c.SemanticModel.GetDeclaredSymbol(propertyDeclaration);
+
+                if (propertySymbol != null &&
+                    IsPubliclyAccessible(propertySymbol) &&
+                    !propertySymbol.IsOverride)
+                {
+                    ReportIfListT(propertyDeclaration.Type, c, "property");
+                }
+            },
+            SyntaxKind.PropertyDeclaration);
+
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+            {
+                if (c.IsTest())
+                {
+                    return;
+                }
+
+                var fieldDeclaration = (FieldDeclarationSyntax)c.Node;
+
+                var variableDeclaration = fieldDeclaration.Declaration?.Variables.FirstOrDefault();
+                if (variableDeclaration == null)
+                {
+                    return;
+                }
+
+                var fieldSymbol = c.SemanticModel.GetDeclaredSymbol(variableDeclaration);
+
+                if (fieldSymbol != null &&
+                    IsPubliclyAccessible(fieldSymbol) &&
```

Comment:
yes, but :)

This looks suspicious to me without checking for null first: fieldSymbol.IsPubliclyAccessible()
SymbolHelper.IsPubliclyAccessible(fieldSymbol) is another option, but I think I'll go for null check + extension method